### PR TITLE
[TT-Train] Add generic inference example

### PR DESCRIPTION
**Pull request by abustamanteTT**
### Problem description
Previous TT-Train overloaded the functionality of nanoGPT example to include training and inference of arbitrary models.

### What's changed
This PR adds a generic LLM inference example in both C++ and Python, which allows users to perform inference with any arbitrary (currently-supported) LLM

### Checklist
https://github.com/tenstorrent/tt-metal/actions/runs/18236304518
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: tt-train/sources/examples/llm_inference/llm_inference.ipynb [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+    "\n",
+    "if padded_vocab_size != vocab_size:\n",
+    "    logits_mask = np.zeros((1, 1, 1, padded_vocab_size), dtype=np.float32)\n",
+    "    logits_mask[:, :, :, vocab_size:] = 1e4\n",
+    "\n",
+    "    logits_mask_tensor = ttml.autograd.Tensor.from_numpy(logits_mask, ttml.Layout.ROW_MAJOR, ttml.autograd.DataType.BFLOAT16)   # [1,1,1,T], float32\n",
+    "else:\n",
+    "    logits_mask_tensor = None\n",
+    "\n",
+    "prompt_str = \"The difference between cats and dogs is:\"\n",
```
Comment (abustamanteTT) [inline tt-train/sources/examples/llm_inference/llm_inference.ipynb:838]: This prompt is hard-coded to allow rapid iteration/debugging; will replace with user input

File: tt-train/sources/examples/llm_inference/llm_inference.ipynb [hunk 2]
```diff
@@ -0,0 +1,736 @@
...
+    "from huggingface_hub import hf_hub_download\n",
+    "from transformers import AutoTokenizer\n",
+    "from yaml import safe_load, Loader\n",
+    "from pathlib import Path\n",
+    "\n",
+    "sys.path.append(f\"{os.environ['TT_METAL_HOME']}/tt-train/sources/ttml\")\n",
+    "import ttml\n",
+    "\n",
+    "def set_seed(seed: int = 42):\n",
+    "    random.seed(seed)\n",
```
Comment (dmakoviichuk-tt) [inline tt-train/sources/examples/llm_inference/llm_inference.ipynb]: you can set seed to our autocontext here too.,

File: tt-train/sources/examples/llm_inference/llm_inference.ipynb [hunk 3]
```diff
@@ -0,0 +1,736 @@
...
+    "    np.random.seed(seed)\n",
+    "\n",
+    "set_seed()\n",
+    "# Change working directory to TT_METAL_HOME\n",
+    "os.chdir(os.environ['TT_METAL_HOME'])\n",
```
Comment (dmakoviichuk-tt) [inline tt-train/sources/examples/llm_inference/llm_inference.ipynb]: lets remove it. Or working dir by default should be tt-train

File: tt-train/sources/examples/llm_inference/llm_inference.ipynb [hunk 4]
```diff
@@ -0,0 +1,736 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "id": "ea71a9df",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "tokenizer = AutoTokenizer.from_pretrained(\"gpt2\")\n",
+    "transformer_cfg = safe_load(open(\"tt-train/configs/training_shakespeare_gpt2s.yaml\", \"r\"))[\"training_config\"][\"transformer_config\"]"
```
Comment (dmakoviichuk-tt) [inline tt-train/sources/examples/llm_inference/llm_inference.ipynb]: lets remove "tt-train" from path

File: tt-train/sources/examples/llm_inference/llm_inference.ipynb [hunk 5]
```diff
@@ -0,0 +1,736 @@
...
+      "Warning: vocab size 50257 is not multiple of 32, padding for tilizing.\n",
+      "Transformer configuration:\n",
+      "    Vocab size: 50272\n",
+      "    Max sequence length: 1024\n",
+      "    Embedding dim: 768\n",
+      "    Num heads: 12\n",
+      "    Dropout probability: 0.2\n",
+      "    Num blocks: 12\n",
+      "    Positional embedding type: Trainable\n",
+      "    Runner type: Default\n",
```
Comment (dmakoviichuk-tt) [inline tt-train/sources/examples/llm_inference/llm_inference.ipynb:159]: should be ok to use default runner type it impacts backward pass only

**Review by dmakoviichuk-tt — APPROVED** [review]
Please cleanup your jupyter notebook before merge a nd fix a few comments.

File: tt-train/sources/examples/llm_inference/main.cpp [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+    return (*model)(data, mask);
+}
+
+int main(int argc, char **argv) {
+    CLI::App app{"GPT2S Inference Example"};
+    argv = app.ensure_utf8(argv);
+
+    std::string model_path = "configs/training_shakespeare_gpt2s";
+    std::string tokenizer_path =
+        "/home/ubuntu/.cache/huggingface/hub/models--gpt2/snapshots/607a30d783dfa663caf39e06633721c8d4cfcd7e/"
```
Comment (dmakoviichuk-tt) [inline tt-train/sources/examples/llm_inference/main.cpp]: dont forget to remove hardcoded path.

File: tt-train/sources/examples/llm_inference/main.cpp [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+    auto logits_padding_mask = ttnn::concat(logits_padding_mask_vector, 3);
+    auto logits_padding_mask_autograd = ttml::autograd::create_tensor(logits_padding_mask);
+
+    for (uint32_t token_idx = 0; token_idx < tokens_to_generate; ++token_idx) {
+        uint32_t start_idx = 0;
+        if (prompt_tokens.size() > max_sequence_length) {
+            start_idx = static_cast<uint32_t>(prompt_tokens.size() - max_sequence_length);
+        }
+        // Fill padded array
+        for (uint32_t i = 0; i < max_sequence_length; ++i) {
```
Comment (athompsonTT) [inline tt-train/sources/examples/llm_inference/main.cpp]: `std::fill(prompt_tokens_padded.begin(), std::advance(prompt_tokens_padded.begin(), max_sequence_length), 0);` perhaps

Comment (abustamanteTT) [reply tt-train/sources/examples/llm_inference/main.cpp]: I think even a `std::fill(prompt_tokens_padded.begin(), prompt_tokens_padded.end(), 0);`?

File: tt-train/sources/examples/llm_inference/main.cpp [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+        for (uint32_t j = 0; j < max_sequence_length; ++j) {
+            mask.push_back(i >= j ? 1.0F : 0.0F);
+        }
+    }
+
+    auto causal_mask_tensor = ttml::autograd::create_tensor(
+        ttml::core::from_vector(mask, ttnn::Shape({1, 1, max_sequence_length, max_sequence_length}), device));
+
+    // Create a large negative mask for out-of-vocab logits
+    auto vocab_mask = std::vector<float>(padded_vocab_size - original_vocab_size, 1e4F);
```
Comment (dmakoviichuk-tt) [inline tt-train/sources/examples/llm_inference/main.cpp]: please add check if you need to apply padding_mask

Comment (dmakoviichuk-tt) [reply tt-train/sources/examples/llm_inference/main.cpp]: Same in python

File: tt-train/sources/examples/llm_inference/main.cpp [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+#include "autograd/tensor.hpp"
+#include "models/gpt2.hpp"
+#include "ops/sampling_op.hpp"
+#include "tokenizers/bpe_tokenizer.hpp"
+#include "ttnn_fixed/trivial_ttnn_ops.hpp"
+
+using MeshShape = tt::tt_metal::distributed::MeshShape;
+using Model = std::shared_ptr<ttml::models::BaseTransformer>;
+
+ttml::autograd::TensorPtr run_model(
```
Comment (rfurko-tt) [inline tt-train/sources/examples/llm_inference/main.cpp]: lets remove run_model method, we don't need it anymore

File: tt-train/sources/examples/llm_inference/main.cpp [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+int main(int argc, char **argv) {
+    CLI::App app{"GPT2S Inference Example"};
+    argv = app.ensure_utf8(argv);
+
+    std::string model_path = "configs/training_shakespeare_gpt2s";
+    std::string tokenizer_path = "data/gpt2-tokenizer.json";
+    std::string prompt = "";
+    std::string safetensors_path = "";
+    float temperature = 0.0F;
+    uint32_t tokens_to_generate = 256;
```
Comment (rfurko-tt) [inline tt-train/sources/examples/llm_inference/main.cpp]: 256U

File: tt-train/sources/examples/llm_inference/main.cpp [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+    uint32_t seed = ttml::autograd::ctx().get_generator()();
+
+    app.add_option("-c,--config", model_path, "Path to the LLM config")->default_val(model_path);
+    app.add_option("-p,--prompt", prompt, "Prompt for text generation")->default_val(prompt);
+    app.add_option("--safetensors", safetensors_path, "Path to the model safetensors file")
+        ->default_val(safetensors_path);
+    app.add_option("--tokenizer", tokenizer_path, "Path to the tokenizer files")->default_val(tokenizer_path);
+    app.add_option("--temperature", temperature, "Sampling temperature")->default_val(temperature);
```
Comment (rfurko-tt) [inline tt-train/sources/examples/llm_inference/main.cpp]: isn't it part of the yaml config?

File: tt-train/sources/examples/llm_inference/main.cpp [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+    ttml::autograd::ctx().open_device(MeshShape{1, 1}, {});
+
+    auto *device = &ttml::autograd::ctx().get_device();
+
+    // Load tokenizer
+    auto tokenizer = std::make_shared<ttml::tokenizers::BPETokenizer>(tokenizer_path);
+
+    // Load model
+    auto yaml_config = YAML::LoadFile(model_path + ".yaml");
+    auto config = ttml::models::gpt2::read_config(yaml_config["training_config"]["transformer_config"]);
```
Comment (rfurko-tt) [inline tt-train/sources/examples/llm_inference/main.cpp]: does it only support gpt2 model?

File: tt-train/sources/examples/llm_inference/main.cpp [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+    std::string safetensors_path = "";
+    float temperature = 0.0F;
+    uint32_t tokens_to_generate = 256;
+    uint32_t seed = ttml::autograd::ctx().get_generator()();
+
+    app.add_option("-c,--config", model_path, "Path to the LLM config")->default_val(model_path);
+    app.add_option("-p,--prompt", prompt, "Prompt for text generation")->default_val(prompt);
+    app.add_option("--safetensors", safetensors_path, "Path to the model safetensors file")
+        ->default_val(safetensors_path);
+    app.add_option("--tokenizer", tokenizer_path, "Path to the tokenizer files")->default_val(tokenizer_path);
```
Comment (rfurko-tt) [inline tt-train/sources/examples/llm_inference/main.cpp]: if it's not provided here -> will it use tokenizer from config?

File: tt-train/sources/examples/llm_inference/main.cpp [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+    for (uint32_t i = start_idx; i < prompt_tokens.size(); ++i) {
+        prompt_tokens_padded[i - start_idx] = prompt_tokens[i];
+    }
+
+    // Create input tensor
+    auto prompt_tensor = ttml::autograd::create_tensor(ttml::core::from_vector<uint32_t, ttnn::DataType::UINT32>(
+        prompt_tokens_padded, ttnn::Shape({1U, 1U, 1U, max_sequence_length}), device, ttnn::Layout::ROW_MAJOR));
+
+    // Generate tokens
+    std::cout << "Generated text:\n";
```
Comment (rfurko-tt) [inline tt-train/sources/examples/llm_inference/main.cpp]: my understanding that it answers only a single prompt and closes program. @abustamanteTT  @dmakoviichuk-tt is it expected? 

in `nano_gpt/main.cpp` i was doing while true logic :D

File: tt-train/sources/examples/python/transformers/training.py [hunk 15]
```diff
@@ -13,11 +13,11 @@
 sys.path.append(f'{os.environ["TT_METAL_HOME"]}/tt-train/sources/ttml')
 
 import ttml
+from ttml.common.config import get_config, DeviceConfig, TrainingConfig
```
Comment (rfurko-tt) [inline tt-train/sources/examples/python/transformers/training.py:16]: do we only want to reuse config, nothing else?

File: tt-train/tests/python/test_autograd.py [hunk 16]
```diff
@@ -6,6 +6,7 @@
 import os
 import sys
 
+
 # sys.path.append(f'{os.environ["TT_METAL_HOME"]}/tt-train/build/sources/ttml')
```
Comment (rfurko-tt) [inline tt-train/tests/python/test_autograd.py]: i know it's not part of your PR, but could you please remove comment :)

File: tt-train/sources/examples/llm_inference/main.cpp [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+    std::cout << "*******************\n";
+    std::cout << prompt;
+
+    // Build mask (causal) for attention
+    std::vector<float> mask;
+    mask.reserve(static_cast<size_t>(max_sequence_length * max_sequence_length));
```
Comment (athompsonTT) [inline tt-train/sources/examples/llm_inference/main.cpp:120]: could use std::fill again, not sure if it makes any difference:
`
std::vector<float> mask(max_sequence_length * max_sequence_length);
    for (uint32_t i = 0; i < max_sequence_length; ++i) {
        auto start = mask.begin() + static_cast<size_t>(i * max_sequence_length);
        std::fill(start, start + i, 0.0F);
        std::fill(start + i, start + (max_sequence_length - i), 1.0F);
}
`

Comment (athompsonTT) [reply tt-train/sources/examples/llm_inference/main.cpp:120]: or std::vector<float> mask(max_sequence_length * max_sequence_length, 0.0F); and only do the 2nd fill

File: tt-train/sources/examples/llm_inference/main.cpp (hunk 11 above)
Comment (abustamanteTT) [reply tt-train/sources/examples/llm_inference/main.cpp]: It is, I hadn't wanted to copy code from the nanoGPT example, but I'm noticing that we don't have a unified EvalConfig struct + parsing function. I'll copy those over from nanoGPT, and look into creating a generic config class

File: tt-train/sources/examples/llm_inference/main.cpp (hunk 12 above)
Comment (abustamanteTT) [reply tt-train/sources/examples/llm_inference/main.cpp]: Long term, no, short term I think we only have GPT2 and LLama safetensors

File: tt-train/sources/examples/llm_inference/main.cpp (hunk 14 above)
Comment (abustamanteTT) [reply tt-train/sources/examples/llm_inference/main.cpp]: I wasn't sure if we wanted to do while true, or single prompt 😅

**Review by athompsonTT — APPROVED** [review]

File: tt-train/sources/examples/python/transformers/training.py (hunk 15 above)
Comment (abustamanteTT) [reply tt-train/sources/examples/python/transformers/training.py:16]: Just did a refactor, I think there's more in the other files to reuse, but I'd need to take another look. Commit [544d3da](https://github.com/tenstorrent/tt-metal/pull/29704/commits/544d3dae1c39fb0442c00d2386348248082532aa)

File: tt-train/sources/examples/llm_inference/main.cpp (hunk 13 above)
Comment (abustamanteTT) [reply tt-train/sources/examples/llm_inference/main.cpp]: It will now! Commit [544d3da](https://github.com/tenstorrent/tt-metal/pull/29704/commits/544d3dae1c39fb0442c00d2386348248082532aa)

File: tt-train/sources/examples/llm_inference/main.cpp (hunk 11 above)
Comment (abustamanteTT) [reply tt-train/sources/examples/llm_inference/main.cpp]: Updated with code from nanoGPT in commit [544d3da](https://github.com/tenstorrent/tt-metal/pull/29704/commits/544d3dae1c39fb0442c00d2386348248082532aa), long-term I think a generic config class would be good

File: tt-train/sources/examples/llm_inference/main.cpp (hunk 10 above)
Comment (abustamanteTT) [reply tt-train/sources/examples/llm_inference/main.cpp]: Commit [544d3da](https://github.com/tenstorrent/tt-metal/pull/29704/commits/544d3dae1c39fb0442c00d2386348248082532aa)

File: tt-train/sources/examples/llm_inference/main.cpp (hunk 9 above)
Comment (abustamanteTT) [reply tt-train/sources/examples/llm_inference/main.cpp]: Commit [544d3da](https://github.com/tenstorrent/tt-metal/pull/29704/commits/544d3dae1c39fb0442c00d2386348248082532aa)

File: tt-train/sources/examples/llm_inference/main.cpp (hunk 8 above)
Comment (abustamanteTT) [reply tt-train/sources/examples/llm_inference/main.cpp]: Added check to C++, already existed in Python, commit [544d3da](https://github.com/tenstorrent/tt-metal/pull/29704/commits/544d3dae1c39fb0442c00d2386348248082532aa)

File: tt-train/tests/python/test_autograd.py (hunk 16 above)
Comment (abustamanteTT) [reply tt-train/tests/python/test_autograd.py]: Done!

File: tt-train/sources/examples/llm_inference/main.cpp (hunk 17 above)
Comment (abustamanteTT) [reply tt-train/sources/examples/llm_inference/main.cpp:120]: I like this suggestion, I think it makes it much more readable too. Commit [d2369dd](https://github.com/tenstorrent/tt-metal/pull/29704/commits/d2369dd3893f0c36d7c433108d9dd0e41ac29ed2)

File: tt-train/sources/examples/llm_inference/main.cpp [hunk 18]
```diff
@@ -0,0 +1,243 @@
...
+
+struct TrainingConfig {
+    std::string project_name;
+    std::string model_type;  // one of "gpt2", "llama"
+    uint32_t seed = 5489U;
+    uint32_t model_save_interval = 500;
+    uint32_t batch_size = 64;
+    uint32_t num_epochs = 1;
+    uint32_t max_steps = 5000;
+    float learning_rate = 3e-4F;
```
Comment (dmakoviichuk-tt) [inline tt-train/sources/examples/llm_inference/main.cpp]: I think you can remove everything related to the  training.

File: tt-train/sources/examples/llm_inference/main.cpp (hunk 14 above)
Comment (dmakoviichuk-tt) [reply tt-train/sources/examples/llm_inference/main.cpp]: lets do while true :)

**Review by dmakoviichuk-tt — APPROVED** [review]
we can cleanup inference example a little bit more.
